include/inviwo/core/datastructures/bitset.h

```diff
@@ -0,0 +1,399 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+#pragma once
+
+#include <inviwo/core/common/inviwocoredefine.h>
+#include <inviwo/core/util/exception.h>
+#include <inviwo/core/util/detected.h>
+
+#include <memory>
+#include <vector>
+#include <array>
+#include <iterator>
+#include <type_traits>
+#include <istream>
+#include <ostream>
+
+namespace roaring {
+class Roaring;
+class RoaringSetBitForwardIterator;
+}  // namespace roaring
+
+namespace inviwo {
+
+namespace detail {
+
+template <typename...>
+using void_t = void;
+
+template <class T, class = void>
+struct is_iterator : std::false_type {};
+
+template <class T>
+struct is_iterator<T, void_t<typename std::iterator_traits<T>::iterator_category>>
+    : std::true_type {};
+
+}  // namespace detail
+
+/**
+ * \brief represents a bitset based on roaring bitmaps provided by the CRoaring library
+ */
+class IVW_CORE_API BitSet {
+public:
+    class BitSetIterator {
+    public:
+        using iterator_category = std::bidirectional_iterator_tag;
+        using value_type = uint32_t;
+        using difference_type = uint32_t;
+        using pointer = const uint32_t*;
+        using reference = const uint32_t&;
+
+        BitSetIterator() = default;
+        ~BitSetIterator();
+
+        BitSetIterator& operator++();
+        BitSetIterator operator++(int);
+        BitSetIterator& operator--();
+        BitSetIterator operator--(int);
+
+        value_type operator*() const;
+
+        bool operator==(const BitSetIterator& rhs) const;
+        bool operator!=(const BitSetIterator& rhs) const;
+
+    private:
+        using RoaringIt = roaring::RoaringSetBitForwardIterator;
+
+        friend BitSet;
+        BitSetIterator(const BitSet& b, bool exhausted = false);
+        BitSetIterator(const RoaringIt& it);
+        BitSetIterator(const BitSetIterator& rhs);
+        BitSetIterator& operator=(const BitSetIterator& rhs);
+
+        std::unique_ptr<RoaringIt> it_;
+    };
+
+    BitSet();
+
+    BitSet(const std::vector<uint32_t>& values);
+
+    template <size_t N>
+    BitSet(const std::array<uint32_t, N>& values) : BitSet() {
```

Comment:
Why not use span<uint32_t> constructor instead?

Reply:
because for consecutive data like array and vector one can use `addMany` to add multiple values at once in contrast to adding each value individually.

Reply:
A span is always consecutive. It is the __perfect__ parameter for this case. It provides just the data and size methods you need.